packages/go_router_builder/README.md

```diff
@@ -0,0 +1,314 @@
+    final authorId = int.parse(state.params['authorId']!);
+    return AuthorDetailsScreen(authorId: authorId);
+  },
+),
+```
+
+In this example, the `authorId` parameter is a) required and b) must be an
+`int`. However, neither of these requirements are checked until run-time, making
+it easy to write code that is not type-safe, e.g.
+
+```dart
+void _tap() => context.go('/author/a42'); // error: `a42` is not an `int`
+```
+
+Since Dart is a statically typed language, we'd love to have mistakes get caught
+at compile-time instead of waiting until run-time. The goal of the routing is to
+provide a way to define the required and optional parameters that a specific
+route consumes and to use code generation to take out the drudgery of writing a
+bunch of `go`, `push` and `location` boilerplate code implementations ourselves.
+
+## Defining a route
+
+Define each route as a class extending `GoRouteData` and overriding the `build`
+method.
+
+```dart
+class HomeRoute extends GoRouteData {
+  const HomeRoute();
+
+  @override
+  Widget build(BuildContext context) => HomeScreen(families: familyData);
+}
+```
+
+Required parameters are pulled from the route's `path` defined in the route
+tree.
+
+## Route tree
+
+The tree of routes is defined as an attribute on each of the top-level routes:
+
+```dart
+@TypedGoRoute<HomeRoute>(
+  path: '/',
+  routes: <TypedGoRoute<GoRouteData>>[
+    TypedGoRoute<FamilyRoute>(
+      path: 'family/:familyId',
+    )
+  ],
+)
+class HomeRoute extends GoRouteData {
+  const HomeRoute();
+
+  @override
+  Widget build(BuildContext context) => HomeScreen(families: familyData);
+}
+
+@TypedGoRoute<LoginRoute>(path: '/login')
+class LoginRoute extends GoRouteData {...}
+```
+
+## `GoRouter` initialization
+
+The code generator aggregates all top-level routes into a single list called
+`$appRoutes` for use in initializing the `GoRouter` instance:
+
+```dart
+final _router = GoRouter(routes: $appRoutes);
+```
+
+## Error builder
+
+You can use typed routes to provide an error builder as well:
+
+```dart
+class ErrorRoute extends GoRouteData {
+  ErrorRoute({required this.error});
+  final Exception error;
+
+  @override
+  Widget build(BuildContext context) => ErrorScreen(error: error);
+}
+```
+
+With this in place, you can provide the `errorBuilder` parameter like so:
+
+```dart
+final _router = GoRouter(
+  routes: $appRoutes,
+  errorBuilder: (c, s) => ErrorRoute(s.error!).build(c),
+);
+```
+
+## Navigation
+
+Navigate using the `go` or `push` methods provided by the code generator:
+
+```dart
+void _tap() => PersonRoute(fid: 'f2', pid: 'p1').go(context);
+```
+
+If you get this wrong, the compiler will complain:
+
+```dart
+// error: missing required parameter 'fid'
+void _tap() => PersonRoute(pid: 'p1').go(context);
+```
+
+This, of course, is the whole point of typed routing; the compiler can let us
+known when we get it wrong.
+
+## Query parameters
+
+Optional parameters indicate query parameters:
```

Comment:
This should be called name parameter.

Is there a difference between optional and required name parameters?

also how about optional position parameter. i.e. LoginRoute([this.from]);

Reply:
I forget. Will look.

Reply:
Ah! We don't distinguish between named and positional – it's all about optional!